analyzers/src/SonarAnalyzer.CSharp/Rules/ApiControllersShouldDeriveFromController.cs

```diff
@@ -0,0 +1,90 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ApiControllersShouldDeriveFromController : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6961";
+    private const string MessageFormat = "Inherit from ControllerBase instead of Controller.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    private readonly HashSet<string> viewIdentifiers =
+        [
+           "View",
+           "PartialView",
+           "ViewComponent",
+           "Json",
+           "OnActionExecutionAsync",
+           "OnActionExecuted",
+           "ViewData",
+           "ViewBag",
+           "TempData",
+           "ViewResult"
+        ];
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStartContext =>
+        {
+            if (!compilationStartContext.Compilation.ReferencesControllers())
+            {
+                return;
+            }
+
+            compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+            {
+                var symbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+                if (symbol.IsControllerType()
+                    && symbol.IsPubliclyAccessible()
+                    && IsAnnotatedWithApiAttribute(symbol)
+                    && symbol.BaseType.Is(KnownType.Microsoft_AspNetCore_Mvc_Controller))
+                {
+                    var shouldReportController = true;
+                    symbolStartContext.RegisterSyntaxNodeAction(nodeContext =>
+                    {
+                        if (viewIdentifiers.Contains(nodeContext.Node.GetName()))
+                        {
+                            shouldReportController = false;
+                        }
+                    }, SyntaxKind.IdentifierName);
+
+                    symbolStartContext.RegisterSymbolEndAction(symbolEndContext =>
+                        ReportIssue(symbolEndContext, symbol, shouldReportController));
+                }
+            }, SymbolKind.NamedType);
+        });
+
+    private static void ReportIssue(SonarSymbolReportingContext context, INamedTypeSymbol controllerSymbol, bool shouldReport)
+    {
+        var classDeclaration = (ClassDeclarationSyntax)controllerSymbol.DeclaringSyntaxReferences.First().GetSyntax();
```

Comment:
The code now has a codesmell because of this.
IMO, I do find the new version clearer, and I'd ignore this codesmell as it won't fix with a reasoning.
WDYt?

Reply:
Given that `compilationStartContext` is only used to check for ASP.NET references and to `RegisterSymbolStartAction`, I would get rid of a bit of nesting by extracting the entire lambda of `RegisterSymbolStartAction` into a dedicated `CheckType(ISymbol symbol)` method. WDYT?